### ISSUE TITLE
Detect optionals in String interpolation

### ISSUE THREAD

**Issue by eaigner** [realm/SwiftLint#581]
I don't know if this is possible, but I'd suggest another rule `StringInterpolationOptional`. Consider the following code snippet

```
let world: String? = "world"
let hello = "hello \(world)"
```

This would obviously produce an unexpected result without ever generating a warning or error. The `StringInterpolationOptional` would catch those cases and generate a hard error.

The easiest way would probably be to disallow the `\()` syntax in favor of the `+` operator.

**Comment by scottrhoyt** [comment]
I don't think this produces an unexpected result. This should be the expected result when printing an optional value. I think it would be out of the scope of SwiftLint to try to protect against this sort of logic error, and it would be pretty annoying for those that are expecting this behavior (e.g. debugging `print` statements on optionals).

**Comment by eaigner** [comment]
Why would `hello Optional("world")`be a valid result in any case? I'm not really thinking about `print` as much as I'm about (dangerous) stuff like this

    let url = NSURL(string: "http://example.com?a=\(paramThatCanBeOptional)")

that can go undetected quite some time. Debugging `print` statements should not be in production code anyway and if they are they could be excluded on a per line basis. Or if somebody doesn't want the rule he can just disable it or make it opt-in altogether.

**Comment by scottrhoyt** [comment]
It's a valid result because you are getting the string representation of an optional value. I think it would be an invalid result and introduce more unintended behavior (when an optional is `.None`) to have the string representation of an optional value be the same as a non-optional, but that's just my opinion.

In your example, I would consider it a logic error to treat an optional value as a non-optional, particularly for something as specific as a URL. This is what optional unwrapping and optional binding are for. There are plenty of ways to accomplish this behavior without introducing the possibility of an unintended result.

However, this is just my perspective on the matter. I encourage you to submit a PR with a new rule to implement this behavior as an `OptInRule`. Though, I think it may be difficult to figure out the type information that you will require from what is available via SourceKit.

**Comment by marcelofabri** [comment]
I'm closing this because it'll be in Swift 3.1 (from https://oleb.net/blog/2016/12/optionals-string-interpolation/):

![](https://oleb.net/media/xcode-warning-string-interpolation-optional-as-any.png)

**Comment by jpsim** [comment]
Yeah, that sounds good.
